apps/dot_ext/oauth2_backends.py

```diff
@@ -32,4 +34,22 @@ def create_token_response(self, request):
                 fhir_body["patient"] = cw.fhir_id
                 body = json.dumps(fhir_body)
 
+            # When BENE chooses NOT to share demographic scopes, clean up previous access/refresh tokens.
+            app = token.application
+            user = token.user
+            scope = token.scope
+
+            # Does new token scope NOT contain BENE_PERSONAL_INFO_SCOPES?
+            if not set(settings.BENE_PERSONAL_INFO_SCOPES).issubset(scope.split()):
+                with transaction.atomic():
+                    # Loop thru all AC's for app/user pair.
+                    for tkn in AccessToken.objects.filter(application=app, user=user):
+                        # If not the new token, delete.
+                        if tkn != token:
+                            # Delete related refresh tokens.
+                            for refresh_tkn in RefreshToken.objects.filter(access_token=tkn):
+                                refresh_tkn.delete()
+                            # Delete past token.
+                            tkn.delete()
```

Comment:
Wouldn't it be possible to delete all old tokens in in one statement (reducing database roundtrips) by adding a filter like, `id != token.id` ?  I see the dependency for refresh tokens so that might make it hard, but at least you could batch refresh token deletes by calling `RefreshToken.objects.filter(access_token=tkn).delete()` rather than on each object individually.

Reply:
As we discussed I think these can be moved to single delete calls to the database.  This will reduce latency in testing situations on sandbox where an application may be using a single user to run automated tests which accumulate hundreds if not thousands of tokens.

Reply:
@njdister Great advice for improving the code and performance!  I just replaced that code with two Django queryset filters.   In local testing, removing 10,000 access and refresh tokens took 102 seconds before.  It now takes 52 seconds :-) 

Will push that up shortly.